--- a/drivers/net/ethernet/ibm/emac/core.c	2017-05-05 12:25:44.000000000 +0200
+++ b/drivers/net/ethernet/ibm/emac/core.c	2017-05-05 12:12:11.000000000 +0200
@@ -2678,7 +2678,10 @@
 
 	return 0;
 }
-
+#ifdef CONFIG_NET_POLL_CONTROLLER
+static void emac_netpoll(struct net_device *dev){
+}
+#endif
 static const struct net_device_ops emac_netdev_ops = {
 	.ndo_open		= emac_open,
 	.ndo_stop		= emac_close,
@@ -2690,6 +2693,9 @@
 	.ndo_set_mac_address	= eth_mac_addr,
 	.ndo_start_xmit		= emac_start_xmit,
 	.ndo_change_mtu		= eth_change_mtu,
+#ifdef CONFIG_NET_POLL_CONTROLLER
+	.ndo_poll_controller     = emac_netpoll,
+#endif
 };
 
 static const struct net_device_ops emac_gige_netdev_ops = {
@@ -2703,6 +2709,9 @@
 	.ndo_set_mac_address	= eth_mac_addr,
 	.ndo_start_xmit		= emac_start_xmit_sg,
 	.ndo_change_mtu		= emac_change_mtu,
+#ifdef CONFIG_NET_POLL_CONTROLLER
+	.ndo_poll_controller     = emac_netpoll,
+#endif
 };
 
 static int emac_probe(struct platform_device *ofdev)
